File: app/src/main/java/com/timsimonhughes/atlas/ui/news/NewsAdapter.java
package com.timsimonhughes.atlas.ui.news;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.timsimonhughes.atlas.R;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.ViewHolder> {

    private static final int count = 20;
    private Context context;
    private int[] imgList = {
            R.drawable.example_image,
            R.drawable.example_image_2,
            R.drawable.example_image_2,
            R.drawable.example_image,
            R.drawable.example_image,
            R.drawable.example_image_2,
            R.drawable.example_image_2,
            R.drawable.example_image};

    private String[] titles = {
            "This is a short title",
            "This is a longer title, hence why it's goes on and on like this",
            "This is a short title",
            "This is a short title",
            "This is a longer title, hence why it's goes on and on like this",
            "This is a longer title, hence why it's goes on and on like this",
            "This is a longer title, hence why it's goes on and on like this",
            "This is a short title"
    };

    public NewsAdapter(Context context) {
        this.context = context;
    }

    @NonNull
    @Override
    public NewsAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        View view = layoutInflater.inflate(R.layout.list_item_news, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsAdapter.ViewHolder holder, int position) {
        holder.mRoundedImageView.setImageResource(imgList[position]);
        holder.mTextView.setText(titles[position]);
    }

    @Override
    public int getItemCount() {
        return imgList.length;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        ImageView mRoundedImageView;
        TextView mTextView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            mRoundedImageView = itemView.findViewById(R.id.image_view_news);
            mTextView = itemView.findViewById(R.id.text_view_news_title);

        }
    }
}

File: news/src/main/java/com/timsimonhughes/news/ui/adapter/NewsAdapter.kt
package com.timsimonhughes.news.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.timsimonhughes.news.databinding.ListItemNewsBinding
import com.timsimonhughes.news.model.News

class NewsAdapter(private val clickListener: NewsItemClickListener) :
    ListAdapter<News, NewsViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
        return NewsViewHolder(
            ListItemNewsBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        holder.bind(getItem(position), clickListener)
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<News>() {
            override fun areItemsTheSame(oldItem: News, newItem: News) =
                oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: News, newItem: News) =
                oldItem.title == newItem.title
        }
    }

    interface NewsItemClickListener {
        fun onItemClick(sharedView: View, newsItem: News)
    }
}

class NewsViewHolder(
    private val binding: ListItemNewsBinding
) : RecyclerView.ViewHolder(binding.root) {

    fun bind(newsItem: News, onClick: NewsAdapter.NewsItemClickListener) {
        binding.run {
            this.item = newsItem
            clickListener = onClick
            executePendingBindings()
        }
    }
}